pkg/kubelet/pleg/evented.go

```diff
@@ -0,0 +1,384 @@
+		e.stopCacheUpdateCh = make(chan struct{})
+		go wait.Until(e.watchEventsChannel, 0, e.stopCh)
+		go wait.Until(e.updateGlobalCache, 5*time.Second, e.stopCacheUpdateCh)
+	}
+}
+
+// Stop stops the Evented PLEG
+func (e *EventedPLEG) Stop() {
+	e.runningMu.Lock()
+	defer e.runningMu.Unlock()
+	if isEventedPLEGInUse {
+		close(e.stopCh)
+		close(e.stopCacheUpdateCh)
+		isEventedPLEGInUse = false
+	}
+}
+
+// In case the Evented PLEG experiences undetectable issues in the underlying
+// GRPC connection there is a remote chance the pod might get stuck in a
+// given state while it has progressed in it's life cycle. This function will be
+// called periodically to update the global timestamp of the cache so that those
+// pods stuck at GetNewerThan in pod workers will get unstuck.
+func (e *EventedPLEG) updateGlobalCache() {
+	e.cache.UpdateTime(time.Now())
+}
+
+// Update the relisting period and threshold
+func (e *EventedPLEG) Update(relistingPeriod time.Duration, relistThreshold time.Duration) {
+	e.genericPleg.Update(relistingPeriod, relistThreshold)
+}
+
+// Healthy check if PLEG work properly.
+func (e *EventedPLEG) Healthy() (bool, error) {
+	// GenericPLEG is declared unhealthy when relisting time is more than
+	// the relistThreshold. In case, EventedPLEG is turned on the value of
+	// relistingPeriod, relistingThreshold is adjusted to higher values and
+	// so health of Generic PLEG should take care of the condition checking
+	// higher relistThreshold between last two relist.
+
+	// EventedPLEG is declared unhealthy only if eventChannel is out of capacity.
+	if len(e.eventChannel) >= cap(e.eventChannel) {
+		return false, fmt.Errorf("EventedPLEG: pleg event channel capacity is full with %v events", len(e.eventChannel))
+	}
+
+	timestamp := e.clock.Now()
+	metrics.PLEGLastSeen.Set(float64(timestamp.Unix()))
+	return true, nil
+}
+
+func (e *EventedPLEG) watchEventsChannel() {
+	containerEventsResponseCh := make(chan *runtimeapi.ContainerEventResponse, cap(e.eventChannel))
+	defer close(containerEventsResponseCh)
+
+	// Get the container events from the runtime.
+	go func() {
+		numAttempts := 0
+		for {
+			if numAttempts >= e.eventedPlegMaxStreamRetries {
+				if isEventedPLEGInUse {
+					// Fall back to Generic PLEG relisting since Evented PLEG is not working.
+					e.Stop()
+					e.genericPleg.Stop()                                              // Stop the existing Generic PLEG which runs with longer relisting period when Evented PLEG is in use.
+					e.Update(e.genericPlegRelistPeriod, e.genericPlegRelistThreshold) // Update the relisting period to the default value for the Generic PLEG.
+					e.genericPleg.Start()
+					break
+				}
+			}
+
+			err := e.runtimeService.GetContainerEvents(containerEventsResponseCh)
+			if err != nil {
+				numAttempts++
+				e.Relist() // Force a relist to get the latest container and pods running metric.
+				klog.V(4).InfoS("Evented PLEG: Failed to get container events, retrying:", "err", err)
```

Comment:
should this be a warning?

Reply:
Yes, thanks.

Reply:
I tried replacing `InfoS` with `Warning`, but it results in following compilation error,
```
pkg/kubelet/pleg/evented.go:184:15: klog.V(4).Warning undefined (type klog.Verbose has no field or method Warning)
```

Reply:
Replaced it with `klog.Warningf` and then it compiles.